File: haskell/glue/src/Glue/Module/Registration.hs
module Glue.Module.Registration where

import Control.Monad (foldM)
import Data.Map.Strict qualified as Map
import Data.Text (Text)
import Data.Text qualified as T
import Glue.IR (IR (..))
import Glue.Module (ModuleInfo (..), RegisteredModule (..))
import Glue.Module.Error (ModuleRegistryError (..))
import Glue.Module.Registry (ModuleRegistry)
import Prelude hiding (mod)

-- | Parse a single module from IR
parseModule :: IR m -> Either (ModuleRegistryError m) (ModuleInfo m)
parseModule (List (Symbol "module" : moduleNameIR : body)) = do
    moduleName <- extractModuleName moduleNameIR
    exports <- extractExports body
    definitions <- extractDefinitions body
    pure $ ModuleInfo{..}
parseModule _ = Left $ InvalidModuleStructure "Expected (module name body...)"

-- | Extract module name from Symbol or DottedSymbol
extractModuleName :: IR m -> Either (ModuleRegistryError m) Text
extractModuleName (Symbol name) = pure name
extractModuleName (DottedSymbol parts) = pure $ T.intercalate "." parts
extractModuleName _ = Left $ InvalidModuleStructure "Module name must be a symbol"

-- | Extract exports from module body
extractExports :: [IR m] -> Either (ModuleRegistryError m) [Text]
extractExports [] = pure []
extractExports (List (Symbol "export" : symbols) : rest) = do
    symTexts <- extractSymbolList symbols
    remaining <- extractExports rest
    pure $ symTexts <> remaining
extractExports (_ : rest) = extractExports rest

-- | Extract definitions from module body
extractDefinitions :: [IR m] -> Either (ModuleRegistryError m) [(Text, IR m)]
extractDefinitions [] = pure []
extractDefinitions (List [Symbol "def", Symbol name, val] : rest) = do
    remaining <- extractDefinitions rest
    pure $ (name, val) : remaining
extractDefinitions (_ : rest) = extractDefinitions rest

-- | Extract list of symbols
extractSymbolList :: [IR m] -> Either (ModuleRegistryError m) [Text]
extractSymbolList [] = pure []
extractSymbolList (Symbol s : rest) = do
    remaining <- extractSymbolList rest
    pure $ s : remaining
extractSymbolList invalid = Left $ InvalidExportList invalid

-- | Build registry from multiple modules
buildRegistry :: [IR m] -> Either (ModuleRegistryError m) (ModuleRegistry m)
buildRegistry modules = do
    moduleInfos <- mapM parseModule modules
    foldM addModuleToRegistry Map.empty moduleInfos
  where
    addModuleToRegistry reg info
        | Map.member info.moduleName reg = Left $ DuplicateModuleName info.moduleName
        | otherwise = pure $ Map.insert info.moduleName (moduleInfoToModule info) reg

-- | Convert ModuleInfo to Module
moduleInfoToModule :: ModuleInfo m -> RegisteredModule m
moduleInfoToModule info =
    RegisteredModule
        { name = info.moduleName
        , exports = info.exports
        , body = map (\(sym, val) -> List [Symbol "def", Symbol sym, val]) info.definitions
        }

-- | Register a single module into an existing registry (pure)
registerModule :: ModuleRegistry m -> ModuleInfo m -> Either (ModuleRegistryError m) (ModuleRegistry m)
registerModule registry moduleInfo = do
    let mod = moduleInfoToModule moduleInfo
    if Map.member moduleInfo.moduleName registry
        then Left $ DuplicateModuleName moduleInfo.moduleName
        else Right $ Map.insert moduleInfo.moduleName mod registry

-- | Register multiple modules into a registry (pure)
registerModules :: ModuleRegistry m -> [ModuleInfo m] -> Either (ModuleRegistryError m) (ModuleRegistry m)
registerModules = foldM registerModule

File: dart/glue/lib/src/module/registration.dart
import 'package:glue/either.dart';
import 'package:glue/src/ir.dart';
import 'package:glue/src/module.dart';
import 'package:glue/src/module/registry.dart';

/// Module registration system
/// Parses (module name (export ...) body...) IR structures into RegisteredModule instances

/// Parse a module IR structure into a RegisteredModule
Either<String, RegisteredModule> parseModule(Ir ir) {
  // Must be a list starting with 'module'
  if (ir is! IrList || ir.elements.isEmpty) {
    return Left('Module declaration must be a non-empty list');
  }

  final elements = ir.elements;
  if (elements[0] is! IrSymbol || (elements[0] as IrSymbol).value != 'module') {
    return Left('Module declaration must start with "module"');
  }

  if (elements.length < 3) {
    return Left('Module declaration requires name, exports, and body');
  }

  // Extract module name
  final nameIr = elements[1];
  if (nameIr is! IrSymbol) {
    return Left('Module name must be a symbol');
  }
  final moduleName = nameIr.value;

  // Extract exports
  final exportsIr = elements[2];

  switch (_parseExports(exportsIr)) {
    case Right(value: final exports):
      {
        final body = elements.sublist(3).toList();
        return Right(
          RegisteredModule(name: moduleName, exports: exports, body: body),
        );
      }
    case Left(value: final err):
      return Left(err);
  }
}

/// Parse export list from (export symbol ...)
Either<String, List<String>> _parseExports(Ir ir) {
  if (ir is! IrList) {
    return Left('Export declaration must be a list');
  }

  if (ir.elements.isEmpty) {
    return Left('Export declaration cannot be empty');
  }

  final first = ir.elements[0];
  if (first is! IrSymbol || first.value != 'export') {
    return Left('Export declaration must start with "export"');
  }

  final exports = <String>[];
  for (final element in ir.elements.sublist(1)) {
    if (element is! IrSymbol) {
      return Left('Export list can only contain symbols');
    }
    exports.add(element.value);
  }

  return Right(exports);
}

/// Build registry from multiple module IR structures
Either<String, ModuleRegistry> buildRegistry(List<Ir> moduleIRs) {
  var registry = emptyRegistry();
  for (final ir in moduleIRs) {
    switch (parseModule(ir)) {
      case Right(value: final module):
        {
          switch (registerModule(registry, module)) {
            case Right(value: final newRegistry):
              registry = newRegistry;
            case Left(value: final err):
              return Left(err);
          }
        }
      case Left(value: final err):
        return Left(err);
    }
  }
  return Right(registry);
}

/// Parse and register multiple modules
Either<String, ModuleRegistry> registerModulesFromIR(
  ModuleRegistry registry,
  List<Ir> moduleIRs,
) {
  var currentRegistry = registry;
  for (final ir in moduleIRs) {
    switch (parseModule(ir)) {
      case Right(value: final module):
        {
          switch (registerModule(currentRegistry, module)) {
            case Right(value: final newRegistry):
              currentRegistry = newRegistry;
            case Left(value: final err):
              return Left(err);
          }
        }
      case Left(value: final err):
        return Left(err);
    }
  }
  return Right(currentRegistry);
}
